phylo/src/evolutionary_models/mod.rs

```diff
@@ -61,34 +79,55 @@ impl Display for ProteinModelType {
     }
 }
 
-pub trait EvolutionaryModelParameters<T> {
-    fn new(model_type: &T, params: &[f64]) -> Result<Self>
+impl ProteinModelType {
+    pub fn get_model_type(model_name: &str) -> Self {
+        match model_name.to_uppercase().as_str() {
+            "WAG" => ProteinModelType::WAG,
+            "BLOSUM" => ProteinModelType::BLOSUM,
+            "HIVB" => ProteinModelType::HIVB,
+            _ => {
+                warn!("Unknown DNA model requested, defaulting to WAG.");
+                ProteinModelType::WAG
+            }
+        }
+    }
+}
+
+pub trait EvoModelParams {
+    type ModelType;
+    type Parameter;
+    fn new(model: &Self::ModelType, params: &[f64]) -> Result<Self>
     where
         Self: Sized;
-    fn get_value(&self, param_name: &Parameter) -> f64;
-    fn set_value(&mut self, param_name: &Parameter, value: f64);
+    fn parameter_definition(
+        model_type: &Self::ModelType,
+    ) -> Vec<(&'static str, Vec<Self::Parameter>)>;
+    fn get_value(&self, param_name: &Self::Parameter) -> f64;
+    fn set_value(&mut self, param_name: &Self::Parameter, value: f64);
+    fn get_pi(&self) -> &FreqVector;
     fn set_pi(&mut self, pi: FreqVector);
 }
 
-impl<const N: usize> std::fmt::Debug for dyn EvolutionaryModel<N> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "EvolutionaryModel with {} states", N)
-    }
-}
 // TODO: change pi to a row vector
-pub trait EvolutionaryModel<const N: usize> {
-    fn new(model_type: ModelType, params: &[f64]) -> Result<Self>
+pub trait EvolutionaryModel {
+    type ModelType;
+    type Params;
+    fn new(model: Self::ModelType, params: &[f64]) -> Result<Self>
     where
-        Self: std::marker::Sized;
+        Self: Sized;
     fn get_p(&self, time: f64) -> SubstMatrix;
```

Comment:
ditto: getters should be nouns, not verbs

Reply:
Done.